src/vision/deliverables/async_fn/async_drop.md

```diff
@@ -0,0 +1,115 @@
+# Async drop
+
+## Impact
+
+* Able to create types (database connections etc) that perform async operations on cleanup
+* Able to detect when such types are dropped synchronously
+* Able to identify the await points that result from async cleanup if needed
+
+## Requires
+
+* [inline async fn support](./inline_async_fn.md)
+
+## Design notes
+
+We can create a `AsyncDrop` variant that contains an `async fn`:
+
+```rust
+impl AsyncDrop for MyType {
+    async fn drop(&mut self) {
+        ...
+    }
+}
+```
+
+Like `Drop`, the `AsyncDrop` trait must be implemented for all values of its self-type.
+
+### Async drop glue
+
+Within async functions, when we drop a value, we will invoke "async drop glue" instead of "drop glue". "Async drop glue" works in the same basic way as "drop glue", except that it invokes `AsyncDrop` where appropriate (and may suspend):
+
+* The async drop glue for a type T first executes the `AsyncDrop` method
+    * If `T` has no `AsyncDrop` impl, then the glue executes the synchronous `Drop` impl
+        * If `T` has no `Drop` impl, then this is a no-op
+* The async drop glue then recursively "async drops" all fields of T
+
+### Requires inline fn 
+
+Making this work requires [inline async fn]. This is because Rust presently assumes *all* types are croppable. Consider a function `foo`:
```

Comment:
```suggestion
Making this work requires [inline async fn]. This is because Rust presently assumes *all* types are droppable. Consider a function `foo`:
```